Make diverses default directories more available

levitte opened the pull request:
## Add the possibility to display and use MODULESDIR

This adds the flag OPENSSL_MODULES_DIR for OpenSSL_version(), and the
flag '-m' for 'openssl version'.

##  Add a way for the application to get OpenSSL configuration data

OpenSSL_version(OPENSSL_DIR) gives you a nicely formatted string for
display, but if all you really want is the directory itself, you were
forced to parsed the string.

This introduces a new function to get diverse configuration data from
the library, OPENSSL_info().  This works the same way as
OpenSSL_version(), but has its own series of types, currently
including:

OPENSSL_INFO_CONFIG_DIR         returns OPENSSLDIR
OPENSSL_INFO_ENGINES_DIR        returns ENGINESDIR
OPENSSL_INFO_MODULES_DIR        returns MODULESDIR
OPENSSL_INFO_DSO_EXTENSION      returns DSO_EXTENSION

For scripting purposes, this also adds the command 'openssl info'.

levitte committed:
Add the possibility to display and use MODULESDIR

This adds the flag OPENSSL_MODULES_DIR for OpenSSL_version(), and the
flag '-m' for 'openssl version'.

levitte committed:
Add a way for the application to get OpenSSL configuration data

OpenSSL_version(OPENSSL_DIR) gives you a nicely formatted string for
display, but if all you really want is the directory itself, you were
forced to parsed the string.

This introduces a new function to get diverse configuration data from
the library, OPENSSL_info().  This works the same way as
OpenSSL_version(), but has its own series of types, currently
including:

OPENSSL_INFO_CONFIG_DIR         returns OPENSSLDIR
OPENSSL_INFO_ENGINES_DIR        returns ENGINESDIR
OPENSSL_INFO_MODULES_DIR        returns MODULESDIR
OPENSSL_INFO_DSO_EXTENSION      returns DSO_EXTENSION

For scripting purposes, this also adds the command 'openssl info'.

levitte commented:
This might for example be useful for @slontis' effort with FIPS module installation and testing.

levitte commented:
Hold on, I'm going to rework this one a bit.

levitte commented:
Rework done.

slontis commented on CHANGES:
```diff
@@ -9,6 +9,11 @@
 
  Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
 
+  *) Added OPENSSL_info() to get diverse built in OpenSSL data, such
```
Change to built-in otherwise it reads funny.

levitte replied:
I'll never get the hang of when to and when not to use dashes...  Fixing

levitte replied:
(and we use "built in" all over the place, it seems.  Still fixing my stuff here, though

slontis replied:
The only other thing I can think of is the pesky windows backslashes, when building a path

mattcaswell replied:
> I'll never get the hang of when to and when not to use dashes

Neither will I.

levitte replied:
Ah, directory-file separator.  Sure, and that will fix the same problem for VMS.

levitte replied:
Also, list separator character may be useful (we use LIST_SEPARATOR internally in a few places)

levitte replied:
Okie, both added

levitte committed:
fixup! Add a way for the application to get OpenSSL configuration data

levitte committed:
squash! Add a way for the application to get OpenSSL configuration data

Additionally:

OPENSSL_INFO_DIR_FILENAME_SEPARATOR     returns directory/filename separator
OPENSSL_INFO_LIST_SEPARATOR             returns list separator

levitte committed:
fixup! Add a way for the application to get OpenSSL configuration data

Including the implementation does help a bit, yeah?  ;-)

levitte committed:
fixup! Add a way for the application to get OpenSSL configuration data

richsalz commented:
Use a dash, "built-in" when it is an adjective like "built-in data"

Less helpfully, do not use a dash when it is a verb, as in "built in the factory"

levitte commented:
Rebased because util/libcrypto.num.

Final review, anyone?

levitte commented:
Ping?

mattcaswell commented on doc/man1/info.pod:
```diff
@@ -0,0 +1,76 @@
+
+=head1 DESCRIPTION
+
+This command is used to print out info information about OpenSSL.
```
info information?

levitte replied:
Yes, exactly.  It's info about the information that you will eventually...

...  nah, kidding, will fix.

levitte replied:
I ended up adding a bit more

mattcaswell reviewed (commented):
I'm not really sure why we need a new command line app for this?

levitte committed:
fixup! Add a way for the application to get OpenSSL configuration data

levitte commented:
Ping?

paulidale reviewed (approved):
LGTM.

paulidale commented:
After the conflicts are resolved.

levitte commented:
Merged with ordinal fixup.

0109e030db9207a47e195b4c3a3b13e9017f0ed2 Add a way for the application to get OpenSSL configuration data
47ca8338358b01ef429a3801ce6173f7a0791674 Add the possibility to display and use MODULESDIR
